### PR TITLE
Fix fusing of p2p barrier tasks

### PR DESCRIPTION
**Pull request by phofl**
- [ ] Closes #xxxx
- [ ] Tests added / passed
- [ ] Passes `pre-commit run --all-files`

see https://github.com/dask/distributed/pull/8944

### CODE REVIEW

File: continuous_integration/environment-3.12.yaml [hunk 1]
```diff
@@ -75,5 +75,5 @@ dependencies:
   - matplotlib  # visualize test
   - pip
   - pip:
-    - git+https://github.com/dask/distributed
+    - git+https://github.com/phofl/distributed@barrier-task
```
Comment (phofl) [inline continuous_integration/environment-3.12.yaml:78]:
```suggestion
    - git+https://github.com/dask/distributed
```

TODO before merging

**Comment by phofl** [comment]
I've only linked to the distributed PR in the 3.12 build, so the others will have one test failure

File: dask/_task_spec.py [hunk 2]
```diff
@@ -857,7 +865,11 @@ def _check_data_node(tsk):
         deps = dependencies[key]
         dependents_key = dependents[key]
 
-        if len(deps) != 1 and len(dependents_key) != 1 or _check_data_node(dsk[key]):
+        if (
+            len(deps) != 1
+            and len(dependents_key) != 1
+            or _check_non_fusable_node(dsk[key])
```
Comment (fjetter) [inline dask/_task_spec.py:871]: very, very small nit: I know this is already the old version but generally I recommend to not have a function dispatch in a hot loop like this. that's unnecessary overhead 

adding a comment offer a similar way to communicate this without paying for the function dispatch
```suggestion           
            # check non fusable
            or tsk.block_fusion
            or (isinstance(tsk, DataNode)
            and not isinstance(tsk.value, str))
```

**Review by fjetter — APPROVED** [review]

**Comment by fjetter** [comment]
I haven't checked the CI failures

File: dask/_task_spec.py (hunk 2 above)
Comment (phofl) [reply dask/_task_spec.py:871]: Yeah, I'll rip out the data thing now and then we can inline

**Pull request merged**